File: FCL/src/main/java/com/tungsten/fcl/ui/version/VersionListAdapter.java
package com.tungsten.fcl.ui.version;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.appcompat.widget.AppCompatImageView;

import com.mio.util.AnimUtil;
import com.tungsten.fcl.R;
import com.tungsten.fclcore.mod.ModManager;
import com.tungsten.fclcore.task.Schedulers;
import com.tungsten.fclcore.task.Task;
import com.tungsten.fcllibrary.component.FCLAdapter;
import com.tungsten.fcllibrary.component.theme.ThemeEngine;
import com.tungsten.fcllibrary.component.view.FCLImageButton;
import com.tungsten.fcllibrary.component.view.FCLRadioButton;
import com.tungsten.fcllibrary.component.view.FCLTextView;

import java.util.ArrayList;

public class VersionListAdapter extends FCLAdapter {

    private final ArrayList<VersionListItem> list;

    public VersionListAdapter(Context context, ArrayList<VersionListItem> list) {
        super(context);
        this.list = list;
    }

    static class ViewHolder {
        FCLRadioButton radioButton;
        AppCompatImageView icon;
        FCLTextView title;
        FCLTextView tag;
        FCLTextView subtitle;
        FCLImageButton delete;
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public Object getItem(int i) {
        return list.get(i);
    }

    @SuppressLint("DefaultLocale")
    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        final ViewHolder viewHolder;
        if (view == null) {
            viewHolder = new ViewHolder();
            view = LayoutInflater.from(getContext()).inflate(R.layout.item_version, null);
            viewHolder.radioButton = view.findViewById(R.id.radio);
            viewHolder.icon = view.findViewById(R.id.icon);
            viewHolder.title = view.findViewById(R.id.title);
            viewHolder.tag = view.findViewById(R.id.tag);
            viewHolder.subtitle = view.findViewById(R.id.subtitle);
            viewHolder.delete = view.findViewById(R.id.delete);
            view.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) view.getTag();
        }
        VersionListItem versionListItem = list.get(i);
        viewHolder.radioButton.checkProperty().unbind();
        viewHolder.radioButton.checkProperty().bind(versionListItem.selectedProperty());
        viewHolder.icon.setBackground(versionListItem.getDrawable());
        viewHolder.title.setText(versionListItem.getVersion());
        viewHolder.tag.setVisibility(versionListItem.getTag() == null ? View.GONE : View.VISIBLE);
        if (versionListItem.getTag() != null) {
            viewHolder.tag.setText(versionListItem.getTag());
        }
        viewHolder.subtitle.setText(versionListItem.getLibraries());
        viewHolder.radioButton.setOnClickListener(view1 -> versionListItem.getProfile().setSelectedVersion(versionListItem.getVersion()));
        viewHolder.delete.setOnClickListener(view1 -> Versions.deleteVersion(getContext(), versionListItem.getProfile(), versionListItem.getVersion()));
        viewHolder.subtitle.setTag(i);
        Task.supplyAsync(() -> {
            ModManager modManager = versionListItem.getProfile().getRepository().getModManager(versionListItem.getVersion());
            return modManager.getMods().size();
        }).whenComplete(Schedulers.androidUIThread(), (result, exception) -> {
            int modCount = exception == null ? result : 0;
            if (((int) viewHolder.subtitle.getTag()) == i) {
                viewHolder.subtitle.setText(String.format("%s  Mods:%d", viewHolder.subtitle.getText(), modCount));
            }
        }).start();
        AnimUtil.playTranslationX(view, ThemeEngine.getInstance().getTheme().getAnimationSpeed() * 30L, -100f, 0f).start();
        return view;
    }
}

File: FCL/src/main/java/com/tungsten/fcl/ui/version/VersionListAdapter.kt
package com.tungsten.fcl.ui.version

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.mio.ui.adapter.ViewHolder
import com.mio.util.AnimUtil.Companion.playTranslationX
import com.tungsten.fcl.activity.MainActivity
import com.tungsten.fcl.databinding.ItemVersionBinding
import com.tungsten.fcllibrary.component.theme.ThemeEngine
import java.nio.file.Files
import kotlin.io.path.isRegularFile

class VersionListAdapter(val context: Context, initList: List<VersionListItem>) :
    RecyclerView.Adapter<ViewHolder>() {
    private val versionList = mutableListOf<VersionListItem>()

    init {
        versionList.addAll(initList)
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        return ViewHolder(
            ItemVersionBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            ).root
        )
    }

    @SuppressLint("DefaultLocale")
    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) {
        val versionListItem = versionList[position]
        val binding = ItemVersionBinding.bind(holder.itemView)
        binding.radioButton.checkProperty().unbind()
        binding.radioButton.checkProperty().bind(versionListItem.selectedProperty())
        binding.icon.background = versionListItem.drawable
        binding.title.text = versionListItem.version
        binding.tag.visibility = if (versionListItem.tag == null) View.GONE else View.VISIBLE
        if (versionListItem.tag != null) {
            binding.tag.text = versionListItem.tag
        }
        binding.subtitle.text = versionListItem.libraries
        binding.radioButton.setOnClickListener {
            versionListItem.profile.selectedVersion = versionListItem.version
        }
        binding.delete.setOnClickListener {
            Versions.deleteVersion(
                context,
                versionListItem.profile,
                versionListItem.version
            )
        }
        binding.root.setOnClickListener {
            versionListItem.profile.selectedVersion = versionListItem.version
        }
        if (!versionListItem.profile.getVersionSetting(versionListItem.version).isGlobal) {
            binding.setting.visibility = View.VISIBLE
            binding.setting.setOnClickListener {
                versionListItem.profile.selectedVersion = versionListItem.version
                val uiManager = MainActivity.getInstance().uiManager
                MainActivity.getInstance().binding.manage.isSelected = true
                uiManager.manageUI.runAfterInit {
                    val tab = uiManager.manageUI.tabLayout.getTabAt(0)
                    uiManager.manageUI.tabLayout.selectTab(tab)
                }
            }
        } else {
            binding.setting.visibility = View.GONE
        }
        val modCount =
            runCatching {
                val modDir =
                    versionListItem.profile.repository.getModsDirectory(versionListItem.version)
                Files.list(modDir).filter {
                    it.isRegularFile()
                }.count().toInt()
            }.getOrNull() ?: 0
        binding.subtitle.text = String.format(
            "%s  Mods:%d",
            binding.subtitle.getText(),
            modCount
        )
        playTranslationX(
            binding.root,
            ThemeEngine.getInstance().getTheme().animationSpeed * 30L,
            -100f,
            0f
        ).start()
    }

    override fun getItemCount(): Int {
        return versionList.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun updateVersionList(versionList: List<VersionListItem>) {
        this.versionList.clear()
        this.versionList.addAll(versionList)
        notifyDataSetChanged()
    }
}
